X eventpromo

ludrob opened the pull request:
This is a port of https://github.com/Financial-Times/n-eventpromo to x-dash.

- Eventpromos are manage by live.ft.com, see: https://live.ft.com/Events/2018/FT-Future-of-Retail-Summit
- We use https://github.com/Financial-Times/next-eventpromo-api to sync/get the data from their system
- Currently, eventpromos look like this (look just after the article content): https://www.ft.com/content/b47de80a-9989-11e8-9702-5946bae86e6d
- The data available for the eventpromo can be accessed here: https://www.ft.com/eventpromo/api

Note: we called this 'eventpromo' to prevent any issues with 'event' and make it easily searchable.

i-like-robots commented on jest.config.js:
```diff
@@ -1,4 +1,6 @@
-	testPathIgnorePatterns: ['/node_modules/', '/bower_components/']
+	testPathIgnorePatterns: ['/node_modules/', '/bower_components/'],
+	verbose: true,
+	testURL: "http://localhost/",
```
This change is in a few branches that I've seen, I'm not sure why it isn't in master yet! But thank you for flagging.

i-like-robots commented on components/x-eventpromo/stories/index.js:
```diff
@@ -0,0 +1,35 @@
+		return text('Eventpromo link', data.eventpromoLink);
+	},
+	eventpromoTitle() {
+		return text('Eventpromo title', data.eventpromoTitle);
```
can these props not be named `id`, `link` (or `url`) and `title` rather than requiring the `eventpromo` prefix?

i-like-robots commented on components/x-eventpromo/stories/eventpromo.js:
```diff
@@ -0,0 +1,27 @@
+	'dates',
+	'image1',
+	'image2',
+	'image3',
```
It's not clear to me what the function of the different images is, do they have specific uses?

ludrob replied:
I'll add a bit more details in the PR description.
The eventpromos always have exactly 3 images, used to create and fade in/out animation.
They all have the same function.

i-like-robots commented on components/x-eventpromo/src/Meta.jsx:
```diff
@@ -0,0 +1,12 @@
+const Meta = ({location, dates}) => {
+	return (
+		<div className="event-promo__meta">
+			<p>{location}</p>
```
Should this be an `<address>` element or is it just a word or two?

ludrob replied:
It's just free text, it can be anything. 
Usually it's something like, "London", "New York", "ExCeL London", etc

Thurst-ft replied:
It is likely to just be a place name and City at most but I don't see any harm in using <address>

i-like-robots commented on components/x-eventpromo/src/Image.jsx:
```diff
@@ -0,0 +1,16 @@
+
+	return (
+		<Link url={linkUrl} attrs={{ 'className': 'js-event-promo--animate', 'data-trackable': 'event-promo'}}>
+			<img className="js-event-promo--animate event-promo-inarticle__img fade-0" srcSet={imgSrc} alt="" data-n-image-lazy-load-js="" role="presentation" styles={{size: '(max-width: 46.24em) 100vw, 45vw'}} />
```
Not sure what's going on with the `styles` property here, do you mean `sizes`?

i-like-robots replied:
Because the image has no content for an assistive device to read it may be best to hide it from them by adding `aria-hidden="true"` and `tabindex="-1"` to the link

i-like-robots commented on components/x-eventpromo/src/ImagesContainer.jsx:
```diff
@@ -0,0 +1,16 @@
+
+	return (
+		<div className="event-promo-inarticle__blocks event-promo-inarticle__img-block">
+			<span className="event-promo__control js-event-promo__control" aria-label="control animation"></span>
```
Should this be an interactive element such as a button?

ludrob replied:
it's the play/pause button overlay-ed on top of the animated images, @Thurst-ft , any reason we should not make this a button?

i-like-robots commented on components/x-eventpromo/readme.md:
```diff
@@ -0,0 +1,16 @@
+
+This module also requires [`x-engine`][engine] as a peer dependency. The Engine module is used to inject your chosen runtime into the component. Please read the `x-engine` documentation first if you are consuming `x-` components for the first time in your application.
+
+[engine]: https://github.com/Financial-Times/x-dash/tree/master/packages/x-engine
```
Usage guidelines and a list of accepted properties needs to be added, see [the readme blueprint](https://github.com/Financial-Times/x-dash/blob/master/private/blueprints/component/readme.md) or check x-teaser for reference.

i-like-robots commented on components/x-eventpromo/src/Footer.jsx:
```diff
@@ -0,0 +1,21 @@
+				}}>Register now</Link>
+			</div>
+			<div className="event-promo-inarticle__brand">
+				<span>Presented by</span>
```
I think this `<span>` may be removed

i-like-robots commented on components/x-eventpromo/src/ImagesContainer.jsx:
```diff
@@ -0,0 +1,16 @@
+import { h } from '@financial-times/x-engine';
+import Image from './Image';
+
+const ImagesContainer = ({link, image1, image2, image3}) => {
```
Perhaps this should be an array of image URLs?

i-like-robots commented on components/x-eventpromo/src/Link.jsx:
```diff
@@ -0,0 +1,9 @@
+
+const Link = ({url, attrs = {}, children}) => {
+	return (
+		<a href={url} {...attrs}>{children}</a>
```
Does this need to be a HOC (could components implement the `<a>` anchor element themselves?)

i-like-robots commented on components/x-eventpromo/src/Details.jsx:
```diff
@@ -0,0 +1,23 @@
+const Details = ({link, title, ...props}) => {
+
+	return (
+		<div className="event-promo-inarticle__blocks">
```
I think this needs a heading element (even if hidden and referenced using an `aria-labelledby` attribute) to identify this content

ludrob commented:
I should now have addressed all the comments on the PR.
My react is a bit rusty and my accessibility skills subpar so do not hesitate to educate me :)

Styles are still sitting in the n-eventpromo repo (at least for now).
We need to decide what we do with them (css in js, outside js, keep in separate repo, etc).
Question: is it acceptable to merge this in the current state and deal with styles later?
Note that the class names should all be used either by styles or js.

Thurst-ft commented on components/x-eventpromo/src/Details.jsx:
```diff
@@ -0,0 +1,23 @@
+import { h } from '@financial-times/x-engine';
+import Footer from './Footer';
```
Not sure about this being named 'Footer' but not sure what else to call it

Thurst-ft commented on components/x-eventpromo/src/Image.jsx:
```diff
@@ -0,0 +1,25 @@
+	const imgSrc = `${baseSrc}width=340 340w,${baseSrc}width=400 740w`;
+
+	return (
+		<a href={linkUrl}
```
formattting looks a bit odd

Thurst-ft commented on __tests__/__snapshots__/snapshots.test.js.snap:
```diff
@@ -1,5 +1,131 @@
+      tabIndex="-1"
+    >
+      <img
+        alt=""
```
alt should have event  eventTitle as value

Thurst-ft commented on components/x-eventpromo/src/Image.jsx:
```diff
@@ -0,0 +1,25 @@
+		>
+			<img
+				className="js-event-promo--animate event-promo-inarticle__img fade-0"
+				srcSet={imgSrc} alt=""
```
alt needs a value probably eventTitle and possibly the strapline

i-like-robots replied:
The image is marked as for presentation only so blank alt text is recommended, see https://www.w3.org/WAI/tutorials/images/decorative/

Thurst-ft commented on components/x-eventpromo/stories/eventpromo.js:
```diff
@@ -0,0 +1,27 @@
+	],
+	'link': 'https://live.ft.com/Events/2018/FT-Brexit-and-Beyond-Summit?segmentId=some-segment-id&amp;variant=false',
+	'location': 'planet earth',
+	'title': 'FT Brexit and Beyond Summit'
```
Should we also inlcude the strapline in to this iteration?

i-like-robots commented on components/x-eventpromo/src/Details.jsx:
```diff
@@ -0,0 +1,23 @@
+
+	return (
+		<div className="event-promo-inarticle__blocks" aria-labelledby="details-header">
+			<h1 id="details-header" style={{display: 'none'}}>Event details and information</h1>
```
This isn't the main page headline. Without knowing the context it will sit in it's hard to choose the right level though, but as it's used for labeling rather than a landmark (display none will remove it from the accessibility tree) a `<div>` will probably do, `<div hidden>` would also work.

ludrob commented:
I've now merged everything (style changes) into x-eventpromo, this PR now has everything needed to render the x-eventpromo component.

i-like-robots commented on tools/x-workbench/package.json:
```diff
@@ -5,7 +5,7 @@
   "description": "",
   "main": "register-components.js",
   "scripts": {
-    "start": "start-storybook -p 9001 -c .storybook -s static -h local.ft.com",
```
Please leave this in, there are some components that will fail due to CORS issues without it.

ludrob commented:
@i-like-robots 
I've updated and prettified everything, happy for us to create a pre-release?
@Thurst-ft 
Strapline changes have now been added to this PR

quarterto commented:
From discussion with @ludrob, gonna close this, development is ongoing in `n-eventpromo` and any `x-` version would be ported from that
